### PR TITLE
Blog - Better Together - Earthly + Github Actions

### PR DESCRIPTION
**Pull request by msali123**
Draft

### CODE REVIEW

**Comment by msali123** [comment]
@adamgordonbell @ezeev [Preview Link](https://649595f72741e32d34c35b0e--ephemeral-narwhal-3ac3b2.netlify.app/blog/earthly-github-actions/)

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+### Simplicity
+
+### **Github Actions**
+
+For this project the CI pipeline needs to do 4 things for each package in the monorepo: 1) build the code, 2) run the unit tests, 3) containerize, and 4) push each container to Docker Hub. For Github Actions, this meant writing Dockerfiles for each package and one rather long YAML file to orchestrate everything. In total, that meant writing 160 lines of YAML and 86 lines of Dockerfile. Don't get me wrong, I love YAML for many things, but it does not feel like the best tool for this kind of task (see "[On YAML Discussions](https://earthly.dev/blog/on-yaml-discussions/)").
+
+Also, Github Actions relies on plugins for certain tasks like setting up runtimes or caching. Depending on the plugin, there can be quite a few nuances that you have to learn and consider. The caching plugin, for example, requires you to specify which directories should be cached and how they should be keyed for cache hits. If you configure either of these incorrectly, there may be nothing in the logs that will tell you. For example, here I have the caching plugin configured for the Rust server with **a directory that does not exist**:
+
+~~~
```
Comment (adamgordonbell) [inline blog/_posts/2023-07-17-earthly-github-actions.md:82]: Can we make this yaml, for syntax highlighting?

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+For the sake of conciseness, here is the code required to orchestrate the steps for the Node service only in Github Actions on the left column and Earthly on the right column. For Github Actions, I had to write the YAML workflow and a Dockerfile. For Earthly, it's just a single, concise Earthfile and that orchestrates all of the steps I need (including building the Docker image) with extremely efficient caching.
+
+<table>
+  <tr>
+    <th>GHA YAML</th>
+    <th>EARTHLY CODE</th>
+  </tr>
+  <tr>
+    <td>~~~
```
Comment (adamgordonbell) [inline blog/_posts/2023-07-17-earthly-github-actions.md:109]: These tables look pretty good. Can we have yaml and dockerfile on the code blocks to get some syntax highlighting?

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+In this case, I'm using the relative path of the dependency that needs to be built within the monorepo - but I could also use a path to another Github repository or branch (in a polyrepo scenario). Earthly handles either automatically. This ability to reference targets in other Earthfiles, whether local or remote is what makes Earthly great for composing pipelines for both monorepos and polyrepos.
+
+In the root directory of the monorepo I have the "main" Earthfile where the `+main-pipeline` I referenced earlier is defined. `+main-pipeline` is once again referencing other targets - this time targets from the same Earthfile, which are then referencing the targets of specific packages. All build logic is local to its respective package with the root Earthfile orchestrating the monorepo build. The `BUILD` commands are executed in parallel. This approach allows you to layout your build logic in a way that matches the layout of your repository and Earthly automatically figures out what is cached, what needs to be run and the order of execution.
+
+~~~
```
Comment (adamgordonbell) [inline blog/_posts/2023-07-17-earthly-github-actions.md:224]:
```suggestion
~~~{.dockerfile caption=""}
```

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+### Next Steps
+
+Thank you for taking the time to read our article about Github Actions and Earthly. If you're interested in learning more, please [schedule](https://calendly.com/d/y22-s9m-cqn/earthly-demo) a demo or conversation with us. Or, skip the call and try [Earthly Satellites](https://earthly.dev/earthly-satellites). They are remote build runners that make builds fast with an automatic and instantly available build cache. Builds can be triggered by GitHub Actions (or any CI) as well as from your laptop. And they're super simple to use. Get started by following [the steps in our documentation to self-serve Satellites](https://docs.earthly.dev/earthly-cloud/satellites).
+
+- [Earthly website](https://earthly.dev)
+- [Earthly Github project](https://github.com/earthly/)
+- [The code for this article](https://github.com/earthly/earthly-vs-gha)
+
+## Outside Article Checklist
```
Comment (adamgordonbell) [inline blog/_posts/2023-07-17-earthly-github-actions.md:296]: remove check list

**Review by adamgordonbell — CHANGES_REQUESTED** [review]
I left a few pieces of feedback. Great turn around time!

Can you also try to make a nice header? Maybe no words but just the Github logo and the Earthly logo working together somehow? Use your judgment on what you think would look good.

**Comment by msali123** [comment]
Yes Yes Adam this was just draft version for you guys to see that side by side code thing

**Comment by adamgordonbell** [comment]
@ezeev you'll need to provide author details and a picture for the article bio at the bottom of the page.

**Comment by msali123** [comment]
Option 1
EarthlyCoverImageTemplate

Option 2
EarthlyCoverImageTemplate (1)

**Comment by msali123** [comment]
[Updated draft preview link](https://6495eecee0a65f0708f05e1e--ephemeral-narwhal-3ac3b2.netlify.app/blog/earthly-github-actions/)

**Comment by msali123** [comment]
[Preview Link 1](https://649755f4ce5f6e1b82760ddd--ephemeral-narwhal-3ac3b2.netlify.app/blog/earthly-github-actions/)

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
...onorepos. I wanted the project to be a resource I could use to help explain Earthly to someone brand new to it. At the same time I was having a lot of conversations with developers in and outside of the Earthly community and it seemed that many of them were using Github Actions to run at least some part of their build process. So, I thought it would be a good exercise to set up my test project with Github Actions as well as with Earthly to see how the two compare. I ended up learning a lot about how they performed when it came to build consistency, developer experience, and monorepo support. . I've used Github Actions a lot over the past few years and so have many Earthly users. There are some things that I really love about Github Actions. For one, it has a generous free tier and is tightly integrated with Github (obviously). This makes it an obvious choice when starting out if you're already pushing code to Github. But it definitely has many limitations especially as your builds grow in size and complexity. It's not a problem unique to Github Actions by a long shot. Finding the solution is [why Earthly exists](https://earthly.dev/blog/new-fundings-at-earthly/) in the first place.
```
Comment (thtmnisamnstr) [inline blog/_posts/2023-07-17-earthly-github-actions.md:16]: Change "... monorepo support. . I've used Github Actions ..." to "... monorepo support. I've used Github Actions ..."

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+- Project root/
+- Go_server/ (Go server package)
+- Node_server/ (Node.js server package)
+- Python_server/ (Python server package)
+- Rust_server/ (Rust server package)
+- Quote_client/ (React front-end package)
+- Quote_generator/ (Generates a text file of quotes used by the servers)
+
+And here is what the front-end looks like with everything is running:
```
Comment (thtmnisamnstr) [inline blog/_posts/2023-07-17-earthly-github-actions.md:36]: Change to "And here is what the front-end looks like with everything running:"

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+<div class="wide">
+![app]({{site.images}}{{page.slug}}/app.png)\
+</div>
+
+Let's dive into each of the main aspects of Earthly and see how they compared to Github Actions for this project.
```
Comment (thtmnisamnstr) [inline blog/_posts/2023-07-17-earthly-github-actions.md:42]: @ezeev I feel like aspects isn't the right word here. Is it features? benefits?

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+  BUILD ./node_server/+docker --tag=$tag
+~~~
+
+### Speed
+
+### **Github Actions**
+
+Build speed is one of the less subjective dimensions to look at for this comparison. In my tests, for this project, I was able to get builds to complete in about 2 minutes and 15 seconds on Github Actions with caching and concurrency fully configured. Although times fluctuate wildly between as low as 1 minute and 30 seconds to 6 minutes for identical builds. And as mentioned early, occasionally Github Action's Docker caching times-out and fails the build at around 6 minutes.
+
+Developer speed is also an important dimension to consider. As I mentioned earlier in this article, the lag time in between iterations is a drag on developer productivity. And it just takes more lines of YAML to do what Earthly can do in fewer lines. YAM
```
Comment (thtmnisamnstr) [inline blog/_posts/2023-07-17-earthly-github-actions.md:276]: This paragraph and the next paragraph should be joined into a single paragraph.

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+
+L is a very easy language to write in. But there is definitely a learning curve to mastering Github Actions and the nuances of its various plugins. Especially when applied to a more complex pattern like a monorepo.
+
+### **Earthly**
+
+With Earthly, executing the same pipeline locally completes in 7 seconds consistently. With [](https://earthly.dev/earthly-ci)[Earthly Satellites](https://earthly.dev/earthly-satellites), it also completes in 7 seconds. This is another data point that speaks to the reproducibility and consistency that Earthly offers.
```
Comment (thtmnisamnstr) [inline blog/_posts/2023-07-17-earthly-github-actions.md:282]: @ezeev When you mention Earthly Satellites here, it's the first mention of it in the post. Expect nobody to know what it is. You should add "– our managed remote build runners –".

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+<div class="wide">
+![7-secs]({{site.images}}{{page.slug}}/7-secs.png)\
+</div>
+
+When it comes to developer speed. There really is no comparison. Composing my CI pipeline in Earthly felt natural and elegant compared to writing it out in YAML. If you've ever written Dockerfiles or Makefiles, Earthly language is easy to learn. And the ability to run my pipeline locally before shipping to CI meant that I was able to iterate really fast and not make a bunch of commits to test my CI pipelines.
```
Comment (thtmnisamnstr) [inline blog/_posts/2023-07-17-earthly-github-actions.md:288]: Change "... Dockerfiles or Makefiles, Earthly language ..." to "... Dockerfiles or Makefiles, Earthly's language ..."

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+
+### Building and Running the Project
+
+If you're interested in running this build yourself - you can find everything at the [earthly-vs-gha repository on Github](https://github.com/earthly/earthly-vs-gha/). The readme includes instructions on how to run it locally as well as on Github Actions. Please note that this project may evolve over time as Earthly and Github Actions release new features and updates or we see opportunities to make improvements.
+
+### Next Steps
+
+Thank you for taking the time to read our article about Github Actions and Earthly. If you're interested in learning more, please [schedule](https://calendly.com/d/y22-s9m-cqn/earthly-demo) a demo or conversation with us. Or, skip the call and try [Earthly Satellites](https://earthly.dev/earthly-satellites). They are remote build runners that make builds fast with an automatic and instantly available build cache. Builds can be triggered by GitHub Actions (or any CI) as well as from your laptop. And they're super simple to use. Get started by following [the steps in our documentation to self-serve Satellites](https://docs.earthly.dev/earthly-cloud/satellites).
+
+- [Earthly website](https://earthly.dev)
```
Comment (thtmnisamnstr) [inline blog/_posts/2023-07-17-earthly-github-actions.md:310]: Remove these 3 bullet points.

File: blog/_posts/2023-07-17-earthly-github-actions.md [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+author: Evan Pease
+
+internal-links:
+ - Earthly
+ - Github Actions
+ - Github
+---
+
+## Introduction
+
```
Comment (thtmnisamnstr) [inline blog/_posts/2023-07-17-earthly-github-actions.md:15]: @ezeev You should add a 1-2 sentence intro of yourself so readers have context before launching into the article.

**Comment by msali123** [comment]
@ezeev Once you review the suggestions by @thtmnisamnstr then I'll fix the changes.

**Review by ezeev — APPROVED** [review]
LGTM

File: blog/_posts/2023-07-17-earthly-github-actions.md (hunk 5 above)
Comment (ezeev) [reply blog/_posts/2023-07-17-earthly-github-actions.md:16]: LGTM

File: blog/_posts/2023-07-17-earthly-github-actions.md (hunk 6 above)
Comment (ezeev) [reply blog/_posts/2023-07-17-earthly-github-actions.md:36]: LGTM

File: blog/_posts/2023-07-17-earthly-github-actions.md (hunk 9 above)
Comment (ezeev) [reply blog/_posts/2023-07-17-earthly-github-actions.md:282]: Agree, @msali123 - can you add?

File: blog/_posts/2023-07-17-earthly-github-actions.md (hunk 10 above)
Comment (ezeev) [reply blog/_posts/2023-07-17-earthly-github-actions.md:288]: LGTM

File: blog/_posts/2023-07-17-earthly-github-actions.md (hunk 11 above)
Comment (ezeev) [reply blog/_posts/2023-07-17-earthly-github-actions.md:310]: LGTM

File: blog/_posts/2023-07-17-earthly-github-actions.md (hunk 7 above)
Comment (ezeev) [reply blog/_posts/2023-07-17-earthly-github-actions.md:42]: Yeah this is was originally something like "the key value propositions of Earthly" but it was too salesy. How is this?

```
Let's take a look at the project the following lenses: build consistency, simplicity, Monorepo support, and speed. 
```

File: blog/_posts/2023-07-17-earthly-github-actions.md (hunk 12 above)
Comment (thtmnisamnstr) [reply blog/_posts/2023-07-17-earthly-github-actions.md:15]: Maybe mention what you do here. "I joined Earthly at the beginning of 2023. I lead solution architecture. That means I help potential customers figure out how they can best use Earthly and how they would implement it. One major part of my job is creating the demo that we show to potential customers. So when I first started at Earthly, I set out to build a project that I could use to highlight Earthly's strengths as a Continuous Integration (CI) solution..."

Or something like that.

**Comment by adamgordonbell** [comment]
> @ezeev @adamgordonbell The difference b/w the 2 links is the header image if both of these don't looks good, let me know i'll change it.

I think option 2 will work

File: blog/_posts/2023-07-17-earthly-github-actions.md (hunk 12 above)
Comment (ezeev) [reply blog/_posts/2023-07-17-earthly-github-actions.md:15]: How about this?

```
I joined Earthly at the beginning of 2023. I lead solution architecture. One of my primary roles is helping potential customers figure out how they can best leverage Earthly. That often requires building demonstrations that show the "art of the possible". So when I first started at Earthly, I set out to build a project that I could use to highlight Earthly's strengths as a Continuous Integration (CI) solution.
```

Comment (thtmnisamnstr) [reply blog/_posts/2023-07-17-earthly-github-actions.md:15]: I'd drop the "art of the possible part" maybe:
I joined Earthly at the beginning of 2023. I lead solution architecture. One of my primary roles is helping potential customers figure out how they can best leverage Earthly. That often requires building demonstrations that show an example of how they would use Earthly in their projects. So when I first started at Earthly, I set out to build a project that I could use to highlight Earthly's strengths as a Continuous Integration (CI) solution.

Comment (msali123) [reply blog/_posts/2023-07-17-earthly-github-actions.md:15]: Sure I'll do in a while.

**Comment by msali123** [comment]
@ezeev @adamgordonbell @thtmnisamnstr [Preview Link](https://649b4b4f9fced108275be0d5--ephemeral-narwhal-3ac3b2.netlify.app/blog/earthly-github-actions/) after all changes.

**Comment by thtmnisamnstr** [comment]
I'm good with this. @ezeev Any changes you want? If not, merge it, and I'll try to get the banner on the GHA landing page today.

**Review by ezeev — APPROVED** [review]

**Review by adamgordonbell — APPROVED** [review]

**Pull request merged**